### PR TITLE
Ledger inventory

### PR DESCRIPTION
**Pull request by Gilthoniel**
This adds the inventory abstraction to the ledger that will hold the instances.

### CODE REVIEW

File: ledger/inventory/mod.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// Instance is the smallest unit of storage of a ledger.
+type Instance interface {
+	GetKey() []byte
+	GetValue() proto.Message
+}
+
+// Page represents the state of the inventory at the given point in time.
+type Page interface {
+	GetIndex() uint64
+	GetRoot() []byte
```
Comment (nkcr) [inline ledger/inventory/mod.go]: not sure what is the index and the root of a page

File: ledger/inventory/mem/mod.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	if err != nil {
+		return page, xerrors.Errorf("couldn't compute page hash: %v", err)
+	}
+
+	inv.stagingPages[page.root] = page
+
+	return page, nil
+}
+
+// Commit stores the snapshot with the given root permanently to the list of
```
Comment (nkcr) [inline ledger/inventory/mem/mod.go]:
```suggestion
// Commit stores the snapshot with the given id permanently to the list of
```

File: ledger/inventory/mem/mod.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		return sum, xerrors.Errorf("couldn't write the key: %v", err)
+	}
+
+	if i.value == nil {
+		return sum, nil
+	}
+
+	buffer, err := proto.Marshal(i.value)
+	if err != nil {
+		return sum, err
```
Comment (nkcr) [inline ledger/inventory/mem/mod.go]: `NewEncodingError(...)` ?

File: ledger/inventory/mod.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import "github.com/golang/protobuf/proto"
+
+// Instance is the smallest unit of storage of a ledger.
+type Instance interface {
+	GetKey() []byte
+	GetValue() proto.Message
+}
+
+// Page represents the state of the inventory at the given point in time.
```
Comment (nkcr) [inline ledger/inventory/mod.go]:
```suggestion
// Page represents the state of the inventory at some point in time.
```

File: ledger/byzcoin/messages.proto [hunk 5]
```diff
@@ -12,4 +12,5 @@ message TransactionProto {
 // applied.
 message BlockPayload {
     repeated TransactionProto txs = 1;
+    bytes root = 2;
```
[the remark this answers is not in the document]
Comment (nkcr) [reply ledger/byzcoin/messages.proto]: ...afterwards that seems to be the ID of the block.

File: ledger/byzcoin/mod.go [hunk 6]
```diff
@@ -172,6 +175,13 @@ func (ldgr *Ledger) makePayload(txs []Transaction) (*BlockPayload, error) {
 		payload.Txs[i] = packed.(*TransactionProto)
 	}
 
+	snap, err := ldgr.inventory.process(payload)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't process the txs: %v", err)
+	}
+
+	payload.Root = snap.GetRoot()
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply ledger/byzcoin/mod.go]: It's normal, the compiler makes the fields public thus the upper case.

File: ledger/byzcoin/messages.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply ledger/byzcoin/messages.proto]: I renamed it to footprint so it is more clear what it should do.

File: ledger/inventory/mod.go (hunk 1 above)
Comment (Gilthoniel) [reply ledger/inventory/mod.go]: comments added

File: ledger/inventory/mem/mod.go (hunk 2 above)
Comment (Gilthoniel) [reply ledger/inventory/mem/mod.go]: I renamed the parameter to be more consistent.

**Pull request merged**